File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/location/LocationAdapter.java
package com.shevaalex.android.rickmortydatabase.ui.location;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.shevaalex.android.rickmortydatabase.databinding.ItemLocationBinding;
import com.shevaalex.android.rickmortydatabase.models.location.LocationModel;

import me.zhanghai.android.fastscroll.PopupTextProvider;

public class LocationAdapter
        extends PagedListAdapter<LocationModel, LocationAdapter.LocationViewHolder>
        implements PopupTextProvider {
    private final OnLocationClickListener onLocationClickListener;

    LocationAdapter(OnLocationClickListener clickListener) {
        super(DIFF_CALLBACK);
        this.onLocationClickListener = clickListener;
    }

    private static final DiffUtil.ItemCallback<LocationModel> DIFF_CALLBACK
            = new DiffUtil.ItemCallback<LocationModel>() {
        @Override
        public boolean areItemsTheSame(@NonNull LocationModel oldItem, @NonNull LocationModel newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull LocationModel oldItem, @NonNull LocationModel newItem) {
            return newItem.equals(oldItem);
        }
    };

    @NonNull
    @Override
    public LocationViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //instantiate view binding class and pass it to ViewHolder
        ItemLocationBinding binding = ItemLocationBinding
                .inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new LocationViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull LocationViewHolder holder, int position) {
        LocationModel currentLocation = getItem(position);
        if (currentLocation != null) {
            holder.binding.locationNameValue.setText(currentLocation.getName());
            if (holder.binding.locationDimensionValue != null
                    &&  holder.binding.locationTypeValue != null) {
                holder.binding.locationDimensionValue.setText(currentLocation.getDimension());
                holder.binding.locationTypeValue.setText(currentLocation.getType());
            }
        }
    }

    @NonNull
    @Override
    public String getPopupText(int position) {
        LocationModel currentLocation = getItem(position);
        if (currentLocation != null) {
            return currentLocation.getName().substring(0, 1);
        } else {
            return "";
        }
    }

    class LocationViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private final ItemLocationBinding binding;

        LocationViewHolder(ItemLocationBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (getAbsoluteAdapterPosition() != RecyclerView.NO_POSITION) {
                onLocationClickListener.onLocationClick(getAbsoluteAdapterPosition(), v);
            }
        }
    }

    public interface OnLocationClickListener {
        void onLocationClick(int position, View v);
    }
}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/ui/location/list/LocationAdapter.kt
package com.shevaalex.android.rickmortydatabase.ui.location.list

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.card.MaterialCardView
import com.shevaalex.android.rickmortydatabase.R
import com.shevaalex.android.rickmortydatabase.databinding.ItemLocationBinding
import com.shevaalex.android.rickmortydatabase.models.location.LocationModel
import com.shevaalex.android.rickmortydatabase.utils.Constants.Companion.TRANSITION_LOCATION

class LocationAdapter(
        private val locationListener: LocationListener
) : PagedListAdapter<LocationModel, LocationAdapter.LocationViewHolder>(DIFF_CALLBACK) {

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<LocationModel>() {
            override fun areItemsTheSame(oldItem: LocationModel, newItem: LocationModel): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: LocationModel, newItem: LocationModel): Boolean {
                return newItem == oldItem
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LocationViewHolder {
        val itemBind = ItemLocationBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
        )
        return LocationViewHolder(itemBind)
    }

    override fun onBindViewHolder(holder: LocationViewHolder, position: Int) {
        val location = getItem(position)
        location?.let {
            holder.bind(it, locationListener)
        }
    }

    class LocationViewHolder(
            private val itemBind: ItemLocationBinding
    ) : RecyclerView.ViewHolder(itemBind.root) {

        fun bind(location: LocationModel, locationListener: LocationListener) {
            val context = itemBind.root.context
            itemBind.root.setOnClickListener {
                locationListener.onLocationClick(location, itemBind.locationItem)
            }
            itemBind.locationItem.transitionName = TRANSITION_LOCATION.plus(location.id)
            Glide.with(context)
                    .load(location.imageUrl)
                    .override(284, 200)
                    .apply(RequestOptions()
                            .placeholder(R.drawable.location_placeholder)
                            .diskCacheStrategy(DiskCacheStrategy.DATA)
                    )
                    .into(itemBind.locationImage)
            itemBind.locationNameValue.text = location.name
            itemBind.locationDimensionValue?.let {
                it.text = location.dimension
            }
            itemBind.locationTypeValue?.let {
                it.text = location.type
            }
        }

    }

    interface LocationListener {
        fun onLocationClick(location: LocationModel, locationCard: MaterialCardView)
    }

}